twill-core/src/test/java/org/apache/twill/kafka/client/KafkaTest.java

```diff
@@ -170,11 +174,128 @@ public void testKafkaClient() throws Exception {
     Cancellable cancel = kafkaClient.getConsumer().prepare().add(topic, 0, 0).consume(new KafkaConsumer
       .MessageCallback() {
       @Override
-      public void onReceived(Iterator<FetchedMessage> messages) {
+      public long onReceived(Iterator<FetchedMessage> messages) {
+        long nextOffset = Long.MIN_VALUE;
+        while (messages.hasNext()) {
+          FetchedMessage message = messages.next();
+          LOG.info(Charsets.UTF_8.decode(message.getPayload()).toString());
+          latch.countDown();
+        }
+        return nextOffset;
+      }
+
+      @Override
+      public void finished() {
+        stopLatch.countDown();
+      }
+    });
+
+    Assert.assertTrue(latch.await(5, TimeUnit.SECONDS));
+    cancel.cancel();
+    Assert.assertTrue(stopLatch.await(1, TimeUnit.SECONDS));
+  }
+
+  @Test
+  public void testKafkaClientReadFromIdx() throws Exception {
+    String topic = "testClient";
+
+    // Publish 30 messages with indecies the same as offsets within the range 0 - 29
+    Thread t1 = createPublishThread(kafkaClient, topic, Compression.GZIP, "GZIP Testing message", 10);
+    t1.start();
+    t1.join();
+    Thread t2 = createPublishThread(kafkaClient, topic, Compression.NONE, "Testing message", 10, 10);
+    t2.start();
+    t2.join();
+    Thread t3 = createPublishThread(kafkaClient, topic, Compression.SNAPPY, "Snappy Testing message", 10, 20);
+    t3.start();
+    t3.join();
+
+    final int startIdx = 15;
+    final CountDownLatch latch = new CountDownLatch(30 - startIdx);
+    final CountDownLatch stopLatch = new CountDownLatch(1);
+    final BlockingQueue<Long> offsetQueue = new LinkedBlockingQueue<>();
+    // Creater a consumer
+    final SimpleKafkaConsumer consumer = (SimpleKafkaConsumer) kafkaClient.getConsumer();
+    Cancellable initCancel = kafkaClient.getConsumer().prepare().add(topic, 0, 0).consume(new KafkaConsumer
+      .MessageCallback() {
+      long minOffset = -2; // earliest msg
+      long maxOffset = -1; // latest msg
+      @Override
+      // Use binary search to find the offset of the message with the index matching startIdx. Returns the next offset
```

Comment:
Right, since the offsets and indices of the messages are known, it can directly jump to the desired idx without binary search.